In [1]:
from tensorflow.keras.datasets import mnist

#this will be the data for the MNIST test
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [2]:
#reshaping data
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000,28, 28, 1))

#normalize data
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

In [3]:
from tensorflow.keras.utils import to_categorical

#one-hot encodes y values
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [4]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

cnn = Sequential()

#adds first convolution layer
cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu',input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2,2)))

#adds second convolution layer
cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu',input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2,2)))

#flattens results
cnn.add(Flatten())

#only contians one dense layer
cnn.add(Dense(units=10, activation='softmax'))

cnn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 128)       73856     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 128)         0         
_________________________________________________________________
flatten (Flatten)            (None, 3200)              0         
_________________________________________________________________
dense (Dense)                (None, 10)                32010     
Total params: 106,506
Trainable params: 106,506
Non-trainable params: 0
__________________________________________________

In [5]:
import time

cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

startTime = time.time()
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
endTime = time.time()

mnistTimeToComplete = endTime - startTime

Epoch 1/5
844/844 [==============================] - 16s 18ms/step - loss: 0.1690 - accuracy: 0.9494 - val_loss: 0.0531 - val_accuracy: 0.9862
Epoch 2/5
844/844 [==============================] - 15s 18ms/step - loss: 0.0513 - accuracy: 0.9841 - val_loss: 0.0373 - val_accuracy: 0.9883
Epoch 3/5
844/844 [==============================] - 16s 19ms/step - loss: 0.0366 - accuracy: 0.9888 - val_loss: 0.0381 - val_accuracy: 0.9897
Epoch 4/5
844/844 [==============================] - 16s 19ms/step - loss: 0.0262 - accuracy: 0.9916 - val_loss: 0.0347 - val_accuracy: 0.9903
Epoch 5/5
844/844 [==============================] - 16s 19ms/step - loss: 0.0209 - accuracy: 0.9934 - val_loss: 0.0341 - val_accuracy: 0.9900


In [6]:
loss, accuracy = cnn.evaluate(X_test, y_test)

#loss and accuracy for mnist test
print(f'loss: {loss}')
print(f'accuracy: {accuracy}')
print(f'time: {mnistTimeToComplete}')

313/313 [==============================] - 1s 3ms/step - loss: 0.0356 - accuracy: 0.9882
loss: 0.03561628609895706
accuracy: 0.9882000088691711
time: 79.07871413230896


In [7]:
cnn = Sequential()

#adds first convolution layer
cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu',input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2,2)))

#adds second convolution layer
cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu',input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2,2)))

#flattens results
cnn.add(Flatten())

#adds dense layer with 4096 neurons
cnn.add(Dense(units=4096, activation='relu'))

#adds a dense layer
cnn.add(Dense(units=128, activation='relu'))

#adds another dense layer for final ouput
cnn.add(Dense(units=10, activation='softmax'))

cnn.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 11, 11, 128)       73856     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 5, 5, 128)         0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 3200)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 4096)              13111296  
_________________________________________________________________
dense_2 (Dense)              (None, 128)              

In [8]:
import time

cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

startTime = time.time()
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
endTime = time.time()

mnistTimeToComplete = endTime - startTime

Epoch 1/5
844/844 [==============================] - 48s 56ms/step - loss: 0.1109 - accuracy: 0.9649 - val_loss: 0.0404 - val_accuracy: 0.9885
Epoch 2/5
844/844 [==============================] - 51s 61ms/step - loss: 0.0369 - accuracy: 0.9883 - val_loss: 0.0409 - val_accuracy: 0.9883
Epoch 3/5
844/844 [==============================] - 50s 59ms/step - loss: 0.0257 - accuracy: 0.9919 - val_loss: 0.0387 - val_accuracy: 0.9895
Epoch 4/5
844/844 [==============================] - 49s 59ms/step - loss: 0.0180 - accuracy: 0.9943 - val_loss: 0.0335 - val_accuracy: 0.9912
Epoch 5/5
844/844 [==============================] - 48s 57ms/step - loss: 0.0155 - accuracy: 0.9952 - val_loss: 0.0421 - val_accuracy: 0.9908


In [9]:
loss, accuracy = cnn.evaluate(X_test, y_test)

#loss and accuracy for mnist test
print(f'loss: {loss}')
print(f'accuracy: {accuracy}')
print(f'time: {mnistTimeToComplete}')

313/313 [==============================] - 2s 6ms/step - loss: 0.0315 - accuracy: 0.9906
loss: 0.03151347488164902
accuracy: 0.9905999898910522
time: 246.7457139492035


<h1>Results of Layer adjustments</h1>
<h3>removing first dense layer</h3>
<p>loss: 0.03561628609895706<br>
accuracy: 0.9882000088691711<br>
time: 79.07871413230896<p>
<h3>adding 4096 neuron dense layer</h3>
<p>loss: 0.03151347488164902<br>
accuracy: 0.9905999898910522<br>
time: 246.7457139492035<p>
<br>
<p>removing the first dense layer lowers the accuracy by about 2 percent but decreases the time to execute by roughly 20 seconds</p>
<p>adding the 4096 neuron dense layer increases the accuracy by about 1 percent but increases the time to execute by almost 3 times</p>